File: server/src/main/java/ch/ethz/lapis/source/MutationFinder.java
package ch.ethz.lapis.source;

import ch.ethz.lapis.util.ReferenceGenomeData;
import java.util.ArrayList;
import java.util.List;
import org.javatuples.Pair;


public class MutationFinder {

    public static List<MutationNuc> findNucMutations(String alignedSeq) {
        char[] ref = ReferenceGenomeData.getInstance().getNucleotideSequenceArr();
        char[] seq = alignedSeq.toUpperCase().toCharArray();

        // Masking leading and tailing deletions because they are often actually unknowns but appear here as
        // deletions due to aligning.
        for (int i = 0; i < seq.length; i++) {
            if (seq[i] != '-') {
                break;
            }
            seq[i] = 'N';
        }
        for (int i = seq.length - 1; i >= 0; i--) {
            if (seq[i] != '-') {
                break;
            }
            seq[i] = 'N';
        }

        List<MutationNuc> mutations = new ArrayList<>();
        for (int i = 0; i < ref.length; i++) {
            int pos = i + 1;
            char refBase = ref[i];
            char seqBase = seq[i];
            if (seqBase != 'C' && seqBase != 'T' && seqBase != 'A' && seqBase != 'G' && seqBase != '-') {
                continue;
            }
            if (seqBase != refBase) {
                mutations.add(new MutationNuc(pos, Character.toString(seqBase)));
            }
        }
        return mutations;
    }


    public static List<MutationAA> findAAMutations(String gene, String aaSeq) {
        char[] ref = ReferenceGenomeData.getInstance().getGeneAASequencesArr().get(gene);
        char[] seq = aaSeq.toUpperCase().toCharArray();

        List<MutationAA> mutations = new ArrayList<>();
        for (int i = 0; i < ref.length; i++) {
            int pos = i + 1;
            char refBase = ref[i];
            char seqBase = seq[i];
            if (seqBase == 'X') {
                continue;
            }
            if (seqBase != refBase) {
                mutations.add(new MutationAA(gene, pos, Character.toString(seqBase)));
            }
        }
        return mutations;
    }


    /**
     * Everything that is not A, T, C, G or - is considered as unknown.
     */
    public static List<Integer> findNucUnknowns(String aaSeq) {
        char[] seq = aaSeq.toUpperCase().toCharArray();

        // Masking leading and tailing deletions because they are often actually unknowns but appear here as
        // deletions due to aligning.
        for (int i = 0; i < seq.length; i++) {
            if (seq[i] != '-') {
                break;
            }
            seq[i] = 'N';
        }
        for (int i = seq.length - 1; i >= 0; i--) {
            if (seq[i] != '-') {
                break;
            }
            seq[i] = 'N';
        }

        List<Integer> positions = new ArrayList<>();
        for (int i = 0; i < seq.length; i++) {
            if (seq[i] != 'A' && seq[i] != 'T' && seq[i] != 'C' && seq[i] != 'G' && seq[i] != '-') {
                positions.add(i + 1);
            }
        }
        return positions;
    }


    /**
     * X is considered as unknown.
     */
    public static List<Integer> findAAUnknowns(String alignedSeq) {
        char[] seq = alignedSeq.toUpperCase().toCharArray();
        List<Integer> positions = new ArrayList<>();
        for (int i = 0; i < seq.length; i++) {
            if (seq[i] == 'X') {
                positions.add(i + 1);
            }
        }
        return positions;
    }


    /**
     * This takes a list of position integers and introduces range representations if appropriate. Example:
     * 1,5,6,7,8,20 will be transformed to 1,5-8,20
     */
    public static List<String> compressPositionsAsStrings(List<Integer> positions) {
        List<String> result = new ArrayList<>();
        Integer rangeStart = null;
        Integer rangeEnd = null;

        for (int pos : positions) {
            if (rangeStart == null) {
                // Initial case
                rangeStart = pos;
                rangeEnd = pos;
            } else if (pos == rangeEnd + 1) {
                // If the range is being continued
                rangeEnd = pos;
            } else {
                // If the range ended
                if (rangeEnd - rangeStart > 1) {
                    // If there is at least one number in between
                    result.add(rangeStart + "-" + rangeEnd);
                } else if (rangeEnd - rangeStart > 0) {
                    // If there are two different numbers
                    result.add(rangeStart.toString());
                    result.add(rangeEnd.toString());
                } else {
                    // If there is a single number
                    result.add(rangeStart.toString());
                }
                rangeStart = pos;
                rangeEnd = pos;
            }
        }
        if (rangeStart != null) {
            // Finishing up - same code like above
            if (rangeEnd - rangeStart > 1) {
                // If there is at least one number in between
                result.add(rangeStart + "-" + rangeEnd);
            } else if (rangeEnd - rangeStart > 0) {
                // If there are two different numbers
                result.add(rangeStart.toString());
                result.add(rangeEnd.toString());
            } else {
                // If there is a single number
                result.add(rangeStart.toString());
            }
        }
        return result;
    }

    /**
     * This takes a list of position integers and introduces range representations if appropriate. Example:
     * 1,5,6,7,8,20 will be transformed to ([1, 5, 8,  20], [false, true, false, false])
     * The first entry in the tuple are the positions. The second entry in the tuple say whether the corresponding
     * position defines the start of a range.
     */
    public static Pair<short[], boolean[]> compressPositionsAsArrays(List<Short> positions) {
        List<Short> resultPositionsList = new ArrayList<>();
        List<Boolean> resultIsStartRangeList = new ArrayList<>();
        Short rangeStart = null;
        Short rangeEnd = null;

        for (short pos : positions) {
            if (rangeStart == null) {
                // Initial case
                rangeStart = pos;
                rangeEnd = pos;
            } else if (pos == rangeEnd + 1) {
                // If the range is being continued
                rangeEnd = pos;
            } else {
                // If the range ended
                if (rangeEnd - rangeStart > 1) {
                    // If there is at least one number in between
                    resultPositionsList.add(rangeStart);
                    resultPositionsList.add(rangeEnd);
                    resultIsStartRangeList.add(true);
                    resultIsStartRangeList.add(false);
                } else if (rangeEnd - rangeStart > 0) {
                    // If there are two different numbers
                    resultPositionsList.add(rangeStart);
                    resultPositionsList.add(rangeEnd);
                    resultIsStartRangeList.add(false);
                    resultIsStartRangeList.add(false);
                } else {
                    // If there is a single number
                    resultPositionsList.add(rangeStart);
                    resultIsStartRangeList.add(false);
                }
                rangeStart = pos;
                rangeEnd = pos;
            }
        }
        if (rangeStart != null) {
            // Finishing up - same code like above
            if (rangeEnd - rangeStart > 1) {
                // If there is at least one number in between
                resultPositionsList.add(rangeStart);
                resultPositionsList.add(rangeEnd);
                resultIsStartRangeList.add(true);
                resultIsStartRangeList.add(false);
            } else if (rangeEnd - rangeStart > 0) {
                // If there are two different numbers
                resultPositionsList.add(rangeStart);
                resultPositionsList.add(rangeEnd);
                resultIsStartRangeList.add(false);
                resultIsStartRangeList.add(false);
            } else {
                // If there is a single number
                resultPositionsList.add(rangeStart);
                resultIsStartRangeList.add(false);
            }
        }
        short[] resultPositions = new short[resultPositionsList.size()];
        boolean[] resultIsStartRange = new boolean[resultPositionsList.size()];
        for (int j = 0; j < resultPositionsList.size(); j++) {
            resultPositions[j] = resultPositionsList.get(j);
            resultIsStartRange[j] = resultIsStartRangeList.get(j);
        }
        return new Pair<>(resultPositions, resultIsStartRange);
    }
}

File: server_v2/src/base/proc/mutation_finder.rs
use crate::base::constants::NucCode;
use crate::base::constants::NucCode::*;
use bio::data_structures::suffix_array::SuffixArray;
use std::hash::Hash;

#[derive(Clone, Hash, Eq, PartialEq)]
pub struct NucMutation {
    pub position: usize,
    pub to: NucCode,
}

pub fn find_nuc_mutations(mut aligned_seq: Vec<NucCode>, reference: &Vec<NucCode>) -> Vec<NucMutation> {
    // Masking leading and tailing deletions because they are often actually unknowns but appear here as
    // deletions due to aligning.
    for x in aligned_seq.iter_mut() {
        if *x != GAP {
            break;
        }
        *x = N;
    }
    for x in aligned_seq.iter_mut().rev() {
        if *x != GAP {
            break;
        }
        *x = N;
    }

    let mut mutations = Vec::new();
    for (i, ref_code) in reference.iter().enumerate() {
        let seq_code = aligned_seq.get(i).unwrap();
        if *seq_code != C && *seq_code != T && *seq_code != A && *seq_code != G && *seq_code != GAP {
            continue;
        }
        if *seq_code != *ref_code {
            mutations.push(NucMutation {
                position: i + 1,
                to: seq_code.clone(),
            });
        }
    }

    mutations
}

/// Everything that is not A, T, C, G or - is considered as unknown.
pub fn find_nuc_unknowns(mut aligned_seq: Vec<NucCode>) -> Vec<usize> {
    // Masking leading and tailing deletions because they are often actually unknowns but appear here as
    // deletions due to aligning.
    for x in aligned_seq.iter_mut() {
        if *x != GAP {
            break;
        }
        *x = N;
    }
    for x in aligned_seq.iter_mut().rev() {
        if *x != GAP {
            break;
        }
        *x = N;
    }

    let mut positions = Vec::new();
    for (i, seq_code) in aligned_seq.iter().enumerate() {
        if *seq_code != C && *seq_code != T && *seq_code != A && *seq_code != G && *seq_code != GAP {
            positions.push(i + 1);
        }
    }

    positions
}

/// This takes a list of position integers and introduces range representations if appropriate. Example:
/// 1,5,6,7,8,20 will be transformed to 1,5-8,20
pub fn compress_positions_as_strings(positions: &Vec<usize>) -> Vec<String> {
    if positions.is_empty() {
        return Vec::new();
    }

    let mut result = Vec::new();
    let mut range_start = positions.get(0).unwrap();
    let mut range_end = range_start;

    for pos in positions {
        if *pos == range_end {
            // This is the initial case and there is nothing to do as range_start and range_end
            // are already initialized above.
        } else if *pos == range_end + 1 {
            // If the range is being continued
            range_end = *pos;
        } else {
            // If the range ended
            if range_end - range_start > 1 {
                // If there is at least one number in between
                result.push(format!("{}-{}", range_start, range_end));
            } else if range_end - range_start > 0 {
                // If there are two different numbers
                result.push(range_start.to_string());
                result.push(range_end.to_string());
            } else {
                // If there is a single number
                result.push(range_start.to_string());
            }
            range_start = *pos;
            range_end = *pos;
        }
    }
    // Finishing up - same code as above
    if range_end - range_start > 1 {
        // If there is at least one number in between
        result.push(format!("{}-{}", range_start, range_end));
    } else if range_end - range_start > 0 {
        // If there are two different numbers
        result.push(range_start.to_string());
        result.push(range_end.to_string());
    } else {
        // If there is a single number
        result.push(range_start.to_string());
    }

    result
}
